File: sem06/ps/Homework2/PerfumeAppMVVM/PerfumeAppMVVM/Models/Repositories/PerfumeRepository.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

namespace PerfumeAppMVVM.Models.Repositories;

public class PerfumeRepository
{
    private readonly DatabaseContext _context;

    public PerfumeRepository(DatabaseContext context)
    {
        _context = context;
    }

    public async Task<Perfume> CreateAsync(string name, string brand, float price, string description = null,
        int? volumeMl = null, string concentration = null, string gender = null, string imagePath = null)
    {
        var perfume = new Perfume
        {
            Name = name,
            Brand = brand,
            Price = price,
            Description = description,
            VolumeML = volumeMl,
            Concentration = concentration,
            Gender = gender,
            ImagePath = imagePath
        };

        _context.Perfumes.Add(perfume);
        await _context.SaveChangesAsync();
        return perfume;
    }

    public async Task<Perfume> GetByIdAsync(int perfumeId)
    {
        return await _context.Perfumes
            .FirstOrDefaultAsync(p => p.Id == perfumeId);
    }

    public async Task<List<Perfume>> GetAllAsync(int skip = 0, int limit = 100, string sortBy = "Name",
        bool sortDesc = false)
    {
        IQueryable<Perfume> query = _context.Perfumes;

        switch (sortBy)
        {
            case "Name":
                query = sortDesc
                    ? query.OrderByDescending(p => p.Name)
                    : query.OrderBy(p => p.Name);
                break;
            case "Brand":
                query = sortDesc
                    ? query.OrderByDescending(p => p.Brand)
                    : query.OrderBy(p => p.Brand);
                break;
            case "Price":
                query = sortDesc
                    ? query.OrderByDescending(p => p.Price)
                    : query.OrderBy(p => p.Price);
                break;
        }

        return await query
            .Skip(skip)
            .Take(limit)
            .ToListAsync();
    }

    public async Task<List<Perfume>> FilterByBrandAsync(string brand, int skip = 0, int limit = 100)
    {
        return await _context.Perfumes
            .Where(p => EF.Functions.Like(p.Brand, $"%{brand}%"))
            .Skip(skip)
            .Take(limit)
            .ToListAsync();
    }

    public async Task<List<Perfume>> SearchByNameAsync(string name, int skip = 0, int limit = 100)
    {
        return await _context.Perfumes
            .Where(p => EF.Functions.Like(p.Name, $"%{name}%"))
            .Skip(skip)
            .Take(limit)
            .ToListAsync();
    }

    public async Task<Perfume> UpdateAsync(int perfumeId, Dictionary<string, object> updates)
    {
        var perfume = await GetByIdAsync(perfumeId);
        if (perfume == null)
            return null;

        foreach (var update in updates)
        {
            switch (update.Key)
            {
                case "Name":
                    perfume.Name = (string)update.Value;
                    break;
                case "Brand":
                    perfume.Brand = (string)update.Value;
                    break;
                case "Description":
                    perfume.Description = (string)update.Value;
                    break;
                case "Price":
                    perfume.Price = (float)update.Value;
                    break;
                case "VolumeML":
                    perfume.VolumeML = (int?)update.Value;
                    break;
                case "Concentration":
                    perfume.Concentration = (string)update.Value;
                    break;
                case "Gender":
                    perfume.Gender = (string)update.Value;
                    break;
                case "ImagePath":
                    perfume.ImagePath = (string)update.Value;
                    break;
            }
        }

        await _context.SaveChangesAsync();
        return perfume;
    }

    public async Task<bool> DeleteAsync(int perfumeId)
    {
        var perfume = await GetByIdAsync(perfumeId);
        if (perfume == null)
            return false;

        _context.Perfumes.Remove(perfume);
        await _context.SaveChangesAsync();
        return true;
    }
}

File: sem06/ps/Homework3/MVCPerfume/models/repositories/perfume_repository.py
from sqlalchemy.orm import Session
from sqlalchemy import asc, desc
from typing import List, Optional
from models.db_models import Perfume


class PerfumeRepository:
    def __init__(self, db: Session):
        self.db = db

    def create(self, name: str, brand: str, price: float, description: str = None,
               volume_ml: int = None, concentration: str = None, gender: str = None,
               image_path: str = None) -> Perfume:
        db_perfume = Perfume(
            name=name,
            brand=brand,
            price=price,
            description=description,
            volume_ml=volume_ml,
            concentration=concentration,
            gender=gender,
            image_path=image_path
        )
        self.db.add(db_perfume)
        self.db.commit()
        self.db.refresh(db_perfume)
        return db_perfume

    def get_by_id(self, perfume_id: int) -> Optional[Perfume]:
        return self.db.query(Perfume).filter(Perfume.id == perfume_id).first()

    def get_all(self, skip: int = 0, limit: int = 100, sort_by: str = "name", sort_desc: bool = False) -> List[Perfume]:
        query = self.db.query(Perfume)

        if sort_by == "name":
            query = query.order_by(desc(Perfume.name) if sort_desc else asc(Perfume.name))
        elif sort_by == "brand":
            query = query.order_by(desc(Perfume.brand) if sort_desc else asc(Perfume.brand))
        elif sort_by == "price":
            query = query.order_by(desc(Perfume.price) if sort_desc else asc(Perfume.price))

        return query.offset(skip).limit(limit).all()

    def filter_by_brand(self, brand: str, skip: int = 0, limit: int = 100) -> List[Perfume]:
        return (self.db.query(Perfume)
                .filter(Perfume.brand.ilike(f"%{brand}%"))
                .offset(skip).limit(limit).all())

    def search_by_name(self, name: str, skip: int = 0, limit: int = 100) -> List[Perfume]:
        return (self.db.query(Perfume)
                .filter(Perfume.name.ilike(f"%{name}%"))
                .offset(skip).limit(limit).all())

    def update(self, perfume_id: int, **kwargs) -> Optional[Perfume]:
        db_perfume = self.get_by_id(perfume_id)
        if db_perfume:
            for key, value in kwargs.items():
                if hasattr(db_perfume, key):
                    setattr(db_perfume, key, value)
            self.db.commit()
            self.db.refresh(db_perfume)
        return db_perfume

    def delete(self, perfume_id: int) -> bool:
        db_perfume = self.get_by_id(perfume_id)
        if db_perfume:
            self.db.delete(db_perfume)
            self.db.commit()
            return True
        return False
